apps/rewards/src/invite/inviteReward.service.ts

```diff
@@ -0,0 +1,171 @@
+import { KomenciLoggerService } from '@app/komenci-logger'
+import { networkConfig, NetworkConfig } from '@app/utils/config/network.config'
+import { isAccountConsideredVerified } from '@celo/base/lib'
+import { CeloContract, ContractKit } from '@celo/contractkit'
+import { Inject, Injectable } from '@nestjs/common'
+import { Cron, CronExpression } from '@nestjs/schedule'
+import { Raw } from 'typeorm'
+import { v4 as uuidv4 } from 'uuid'
+import { NotifiedBlock } from '../blocks/notifiedBlock.entity'
+import { NotifiedBlockRepository } from '../blocks/notifiedBlock.repository'
+import { InviteReward, RewardStatus } from './inviteReward.entity'
+import { InviteRewardRepository } from './inviteReward.repository'
+
+const WEEKLY_INVITE_LIMIT = 20
+
+@Injectable()
+export class InviteRewardService {
+  isRunning = false
+  komenciAddresses = []
+
+  constructor(
+    private readonly inviteRewardRepository: InviteRewardRepository,
+    private readonly notifiedBlockRepository: NotifiedBlockRepository,
+    private readonly contractKit: ContractKit,
+    @Inject(networkConfig.KEY)
+    private readonly networkCfg: NetworkConfig,
+    private readonly logger: KomenciLoggerService
+  ) {
+    this.komenciAddresses = this.networkCfg.relayers.map(
+      relayer => relayer.externalAccount
+    )
+  }
+
+  @Cron(CronExpression.EVERY_MINUTE)
+  async sendInviteRewards() {
+    if (this.isRunning) {
+      this.logger.log('Skipping because previous run is still ongoing')
+      return
+    }
+    this.isRunning = true
+
+    const cUsdTokenAddress = (
+      await this.contractKit.registry.addressFor(CeloContract.StableToken)
+    ).toLowerCase()
+
+    await this.usingLastNotifiedBlock(async fromBlock => {
+      this.logger.log(
+        `Starting to fetch invite rewards from block ${fromBlock}`
+      )
+      const escrow = await this.contractKit.contracts.getEscrow()
+      const withdrawalEvents = await escrow.getPastEvents('Withdrawal', {
+        fromBlock
+      })
+      this.logger.log(`Withdrawal events received: ${withdrawalEvents.length}`)
+      let maxBlock = fromBlock
+      for (const withdrawal of withdrawalEvents) {
+        const {
+          transactionHash,
+          blockNumber,
+          returnValues: { identifier, token, to }
+        } = withdrawal
+        maxBlock = Math.max(maxBlock, blockNumber)
+
+        const inviter = to.toLowerCase()
+        if (cUsdTokenAddress !== token.toLowerCase()) {
+          continue
+        }
+        const tx = await this.contractKit.web3.eth.getTransaction(
+          transactionHash
+        )
+        if (!this.isKomenciSender(tx.from)) {
+          continue
+        }
+        const invitee = tx.to?.toLowerCase() ?? ''
+        const conditions = await Promise.all([
+          this.addressIsConsideredVerified(inviter, identifier),
+          this.addressIsConsideredVerified(invitee, identifier),
```

Comment:
The inviter and invitee won't have the same phone number identifier

Reply:
That's my bad, I added the inviter check last minute but that obviously doesn't make sense. There's an open PR on Blockscout that adds attestations info: https://github.com/celo-org/blockscout/pull/121 maybe we can merge and use that.

Reply:
I worry a little about adding another dependency which can impact reliability. What if we record the address<>identifier of users during signup in the rewards DB? We'd want to have the ability to back-fill this data but once it's caught up, it will be easy to maintain.

Reply:
I like that idea. In that case though, would it make sense to do something similar to what we're doing here? i.e. monitor the blockchain for verifications? That way we can track all verified people as well, not only those that did it using Komenci.

Reply:
Yeah, makes sense to me.

Reply:
Added it in this PR: https://github.com/celo-org/komenci/pull/167